### ISSUE TITLE
Oauth / OpenIDC authentication does not work for clean logins

### ISSUE THREAD

**Issue by Gummikavalier** [RocketChat/Rocket.Chat#29039]
### Description:

Custom Oauth configuration fails to finish the authentication. 

This issue started in RC 6.x. and is reproducible for us in RC 6.1.4.

In short any new authentications fail to finish their authentication properly using Open ID Connect. Web client either throws the user out back to the login screen, or stalls at the point where it should be reading the user profile and membership information and tokens.

In situations where the user browser contains existing cache of the web client, the authentication still works. Clean private browser sessions fail.

### Steps to reproduce:

1. Open up Firefox in private mode to have a fresh start
2. Go to login page
3. Login as usual using the custom authentication provider or LDAP

### Expected behavior:

User logs in and everything works.

### Actual behavior:

We've observed two things happening.
1. Sometimes the user is directed back to login screen and they end up in a loop
2. Most often the users get this view (image from Chrome console enabled)
![Screenshot from 2023-04-25 10-41-59](https://user-images.githubusercontent.com/6958942/234208310-d3913927-7188-4f1e-88ac-1635d374d2f8.png)

In Chrome private view the user is often able to proceed from here by pressing F5.
In Firefox F5 does not help and the user is completely stuck at this point.

3. In addition to breaking Oauth/OpenIDC login flow, the issue also manifests itself with instances using LDAP authentication in Firefox private sessions. After you refresh your browser window immediately after arriving on the login page, you'll lose the login fields from the view. At least this issue is reproducible also at https://open.rocket.chat.
If you try login with Firefox private session and don't refresh the window, you may see that login succeeds at first. But after the next F5 refresh the web client stalls and shows only three blinking dots like it did in the Oauth setup.
4. Also SAML has been reported to suffer from the same issue.

### Server Setup Information:

- Version of Rocket.Chat Server: 6.1.4 (Enterprise)
- Operating System: RHEL8
- Deployment Method: tar
- Number of Running Instances: 1
- DB Replicaset Oplog: Yes
- NodeJS Version: 14.21.2
- MongoDB Version: 5.0

### Client Setup Information

- Desktop App or Browser Version: Latest Firefox

### Additional context

Particular thing that may affect whether user ends up seeing the stalled web client or gets thrown back all the way back to the login screen may depend on the `samesite: strict` versus `samesite: none` or CSP settings too. Setting samesite flag to none in web server config _does not_ fix the web client stalling with three pulsing dots in Firefox, though.

We have also observed these errors in the console of Firefox during the login attempt:

```
Uncaught (in promise) TypeError: n.updatedAt.getTime is not a function
    loadFromCache https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    init https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    I https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    l https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    t https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    t https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    <anonymous> https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
```

```
Uncaught (in promise) TypeError: n.updatedAt.getTime is not a function
    loadFromCache https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    init https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    f /client/providers/SettingsProvider.tsx:1
    f /client/providers/SettingsProvider.tsx:1
    fl https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1205
    unstable_runWithPriority https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1241
    xa https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1205
    cl https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1205
    sl https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1205
    R https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1241
    onmessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1241
    e https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1241
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    t https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    e https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1232
    ...
```

```
Uncaught (in promise) undefined 0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js:1515:3274
    onStop https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1515
    _livedata_nosub https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1423
    onMessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1423
    onmessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    forEach self-hosted:203
    forEachCallback https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    onmessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    dispatchEvent https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    _dispatchMessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    _didMessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    onmessage https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    (Async: EventHandlerNonNull)
    websocket https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    _try_next_protocol https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    _didClose https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    onfinish https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    emit https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    onfinish https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    emit https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    onreadystatechange https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    (Async: EventHandlerNonNull)
    <anonymous> https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    XHRLocalObject https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    (Async: setTimeout handler)
    delay https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    XHRLocalObject https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    <anonymous> https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    J https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    (Async: setTimeout handler)
    delay https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    J https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    q https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    k https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    _launchConnection https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1393
    allowConnection https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1395
    c https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    a https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:5
    n https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
    g https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    r https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    t https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:3
    <anonymous> https://server_name_redacted/0951b877b67d5d44c64bfc29ff458b11fdcdeafa.js?meteor_js_resource=true:1636
```

**Comment by Gummikavalier** [comment]
@debdutdeb @dudanogueira I hope your guys can reproduce with CustomOauth config.

**Comment by Gummikavalier** [comment]
The setup is Enterprise and monolithic and the issue was reproducible for us using only one instance on the host.

**Comment by Gummikavalier** [comment]
We also tried with both custom and vanilla and completely off CSP (Content Security Policy) without success.

For the Firefox I think that for the new user sessions the web client app cannot update data in browser local storage, which then eventually succeeds for Chrome using refresh. Or Chrome simply does not care.

But for Firefox the IndexedDB in the local browser storage remains empty. When I compare an old working and the new nonworking sessions of Firefox and the same user account, I can see below differences.

New session Indexed DB:
Empty

New session Local Storage:
![Screenshot from 2023-04-25 11-25-32](https://user-images.githubusercontent.com/6958942/234218905-9be6bbd5-0eed-4a3c-9d08-601e76dd370c.png)

Old session Indexed DB:
Contains localforage entries that are mentioned to be in the local storage in the new session.

Old session Local Storage contains only these:
![Screenshot from 2023-04-25 11-28-18](https://user-images.githubusercontent.com/6958942/234219614-6afac9e3-9216-411e-870c-87cfd6fae5a8.png)

So overall I suspect that on Firefox the login simply fails so hard during the process that the required web client authentication data never forms and gets written into the localforage database.

**Comment by Gummikavalier** [comment]
I also compared the local storage differences of private modes. It seems that _the issue starts on Firefox already when arriving at the login screen_. You also get that `n.updateAt.getTime is not a function` error already at the arrival to the login page in Firefox.

On the Chrome the localforage database gets created under the IndexedDB immediately when you enter the the login page.

This explains the difference in operation between these two browsers.

**Comment by Gummikavalier** [comment]
Then in the Firefox the login data are gathered as the separate localforage entries under the localstorage instead (or part of the filling process). But since there is no localforage to save and search for them under the IndexedDB (for the further use), the web client stalls.

The stalling itself does not seem to have much to do with the authentication part anymore, but more like with the tokens of room subscriptions, permissions etc.

In short, to fix this issue the creation of the IndexedDB entries, particularly the creation of localforage should be ensured early enough.

**Comment by andreaswolf** [comment]
This seeems to be the same issue I described in #29031, just on 5.4, not 6.1. But your stack traces look familiar @Gummikavalier.

One difference: In Firefox, reloading with F5 helps in our case, at least that worked for me locally. We have not yet rolled this version out due to these difficulties, so I cannot tell what other users say.

**Comment by Gummikavalier** [comment]
Issue with the combination Firefox private session and LDAP authentication happens when you get `localforage/public-settings` entry saved under the `Local Storage`.

You can again see the login fields after cleaning this out at:
`F12 -> Storage -> Local Storage -> your RC instance site -> Delete all`.

**Comment by andreaswolf** [comment]
No worries, since you're on a different major release, it's good to have a separate ticket IMO.

One thing I noticed in our instance: the login completely breaks and we get RocketChat console errors "No matching login attempt found" when we trigger the OAuth login via a message from a custom JS:

```js
window.postMessage({
    externalCommand: 'call-custom-oauth-login'
    service: '[our custom auth service]',
});
```

When clicking the button, the login works at least sometimes.

**Comment by svenseeberg** [comment]
It seems we're having the same problem, with SAML instead of OAuth though.

**Comment by Gummikavalier** [comment]
The situation may still live, but I haven't been able to reproduce login issues in Chrome after I changed the samesite=strict to samesite=none in the nginx config.

In nginx:
```
proxy_cookie_path / "/; secure; httpOnly; SameSite=strict";
->
proxy_cookie_path / "/; secure; httpOnly; SameSite=none";
```

All issues with Firefox still remain as described in the first post.

It has been known that introducing samesite flag into Chrome was bound to break some web authentication flows as described here for instance:
https://www.identityserver.com/documentation/saml2p/config-sp/samesitecookies/

Google was supposed to do a staggered release at enforcing the samesite policy for all Chrome users. However nothing seemed to happen for a long time. I wonder if this is finally happening.

**Comment by svenseeberg** [comment]
> The situation may still live, but I haven't been able to reproduce login issues in Chrome after I changed the samesite=strict to samesite=none in the nginx config.

At least in my case (SAML) I cannot confirm that this makes a difference.

**Comment by Gummikavalier** [comment]
@svenseeberg Thanks for the update regarding the SAML issue. :+1: 

I raised the knowledge of the Firefox login screen and three pulsing dots issue to the devs in the RC Open Call yesterday. 

I'm still not sure whether that one is related to the original looping Oauth login issue where you just end up back to login screen endlessly.

**Comment by hipoloco** [comment]
I confirm that this problem also occurs with CAS authentication, both in the web client and the desktop client. However, the authentication works on the Android client.

**Comment by Gummikavalier** [comment]
Thanks, that's good info. :+1:

**Comment by hipoloco** [comment]
At the moment the solution I found to get CAS authentication working again is to rollback to version 6.0.8 (from 6.2.2), unlock the migration in mongo and downgrade the database version to 292 (from 294):

```
use rocketchat
db.migrations.update({_id: 'control'},{$set:{locked:false,version:292}})
```

**Comment by Gummikavalier** [comment]
Similar, but not the same, issue to this: https://github.com/RocketChat/Rocket.Chat/issues/29643
